content/blog/modern-text-features/index.Rmd

```diff
@@ -0,0 +1,370 @@
+            plot.background = element_rect('honeydew', NA))
+  )
+  invisible(dev.off())
+}
+get_results <- function(name) {
+  files <- sort(list.files("figs", pattern = name))
+  files <- files[order(grepl('ragg', files), decreasing = TRUE)]
+  files <- c(
+    grep("macOS", files, value = TRUE),
+    grep("Windows", files, value = TRUE),
+    grep("Linux", files, value = TRUE)
+  )
+  files <- c(files[c(T, F, F)], files[c(F, T, F)], files[c(F, F, T)])
+  cat(paste(paste0("<img src=\"figs/", files, "\" width=\"33%\" style=\"display: inline;\">"), collapse = ""))
+}
+theme_set(theme_void() + theme(panel.background = element_rect('gray95', NA), plot.margin = margin(0, 2, 2, 2)))
+```
+
+## Advanced script support
+
+English, being the lingua franca of programming, has generally dominated everything related to text within programming, ranging from encoding to rendering.
+Because of this, the Latin script, which is used in most of the western world, has been the best (or often only) supported script in many text rendering pipelines.
+This has also been true in the R world where the build-in graphic devices has struggled to display scripts that differed from the standard Latin layout (the Cairo devices on Linux being the exception as we will see).
+This is not a jab at the provided graphic devices.
+ragg has handled text rendering in the exact same way up until now and the work that has gone into changing it has not been trivial.
+Technicalities aside, it is about time (overdue, actually) that the graphic system in R begins to support the display of non-Latin script and becomes more inclusive in terms of which languages can be used.
+It is thus with great joy that I can finally support it in ragg.
+
+### Right-to-Left scripts
+
+To start off we will look at a sample of different scripts that poses a problem due to their direction.
+
+```{r, eval=FALSE}
+arabic_text <- "هذا مكتوب باللغة العربية"
+hebrew_text <- "זה כתוב בעברית"
+sindhi_text <- "هي سنڌيءَ ۾ لکيو ويو آهي"
+
+p <- ggplot() + 
+  geom_text(
+    aes(x = 0, y = 3:1, label = c(arabic_text, hebrew_text, sindhi_text)), 
+    family = "Arial"
+  ) + 
+  expand_limits(y = c(0, 4))
+
+preview_devices(p, "rtl_example")
+```
+
+```{r, echo=FALSE, results='asis'}
+get_results("rtl_example")
+```
+
+Now, if you are not trained in reading any of the three languages above it may be hard to see what is right and what is wrong.
+You may, however, look at how the text in the code is rendered in the browser and compare that to the device rendering. If you do that you can see that the Hebrew script is rendered in the wrong direction for all the non-ragg devices (except Cairo on Linux).
+For the other scripts it is less obvious that the direction is wrong since the text glyphs are fundamentally different due to ligature substitutions.
+Still, by comparing to the browser rendering you can see that the same devices failing on the Hebrew script fail here as well.
```

Comment:
```suggestion
If you're not familiar with the languages above it can be hard to see what is right and what is wrong.
You may, however, look at how the text in the code is rendered in the browser and compare that to the device rendering. If you do that you can see that the Hebrew script is rendered in the wrong direction for all the non-ragg devices (except Cairo on Linux).
For the Arabic and Sindhi it's even harder to see what's wrong because the text looks fundamentally different.
That's because both Arabic and Sindhi rely extensively on text substitution rules and ligatures; the way a letter is written depends critically on what letters it is next to.
Still, by comparing to the browser rendering you can see that the same devices failing on the Hebrew script fail here as well.
```